File: DSA_Level-2/HashMap And Heaps/Check If An Array Can Be Divided Into Pairs Whose Sum Is Divisible By K/CheckIfAnArrayCanBeDividedIntoPairsWhoseSumIsDivisibleByK_Solution.java
import java.util.*;

class Main {
	
	public static void solution(int[] arr, int k){
		HashMap<Integer,Integer> map=new HashMap<>();
		for(int a : arr){
			map.put(a%k,map.getOrDefault(a%k,0)+1);
		}

		int mid=0;
		if(map.size()%2==0){
			mid=map.size()/2;
		}
		else{
			mid=map.size()+1/2;
		}

		int value;
		for(int key : map.keySet()){
			if(key<mid){
				int keyValue=map.get(key);			
				if(key==0){
					if(keyValue%2!=0){
						System.out.println("false");
						return;
					}
				}
				else if(keyValue!=map.getOrDefault(k-key,0)){
					System.out.println("false");
					return;
					//value+=keyValue*map.get(k-key);
				}
			}
		}

		System.out.println("true");

	}

	public static void main(String[] args) {
		Scanner scn = new Scanner(System.in);
		int n = scn.nextInt();
		int[] arr = new int[n];
		for(int i = 0 ; i < n; i++){
			arr[i] = scn.nextInt();
		}
		int k = scn.nextInt();
		solution(arr,k);
	}

}

File: DSA_Level-2/HashMap And Heaps/Check If An Array Can Be Divided Into Pairs Whose Sum Is Divisible By K/CheckIfAnArrayCanBeDividedIntoPairsWhoseSumIsDivisibleByK_Solution.js

function CheckIfArrayPartition(arr,k){
    let map=new Map();

    for(let a of arr){
        let mod=a%k;
        if(map.has(mod)){
            map.set(mod,map.get(mod)+1);
        }
        else{
            map.set(mod,1);
        }
    }

    //console.log(map);

    let mid=0;

    if(map.size%2==0){
        mid=map.size/2;
    }
    else{
        mid=(map.size+1)/2;
    }

    //console.log(mid);

    let i=0;
    while(i<mid){
        if(i==0){
            if(map.get(0)%2==1)
                return false;
        }
        else{
            if(map.get(i)!=map.get(k-i))
                return false;
        }
        ++i;
    }

    return true;
}

console.log(CheckIfArrayPartition([9,7,5,3],6));
console.log(CheckIfArrayPartition([9,7,5,3,1],6));